### PR TITLE
perf(backup): walk only the prefix's subtree in FSObjectStore.List

### PR DESCRIPTION
**Pull request by Adelagric**
**Problem.** `FSObjectStore.List` walked the whole root on every call and filtered afterwards, so its cost was O(all stored files) regardless of prefix. Every caller passes a narrow prefix — `prune` lists one collection prefix per collection and per run, `LatestKey` one collection, and #111's write-once probe a single key — so a backup run on the filesystem backend cost O(collections × all stored files). Surfaced during the #111 review; it is pre-existing `List` behaviour, hence a separate PR.

**Change.** The walk starts at the deepest directory the prefix fully names (`<root>/acme/col` for `acme/col/2024-`); the trailing partial segment remains a name filter. A prefix whose directory does not exist yields `nil`, not an error (same as an empty root today). A prefix that would escape root (`..`) falls back to root, where the key filter matches nothing — the walk never leaves root. The key-prefix filter is still applied to every entry, so results are byte-identical to the full-root walk.

**Test.** `TestFSObjectStoreListPrefixScopedWalk` compares `List` against a full-root walk filtered by prefix across empty, collection, partial-segment, exact-key, missing-directory and `..` prefixes, with a real file planted outside root to prove it is never reached.

No interface or key-layout change; `S3Store.List` already scopes server-side via the `prefix` parameter, `MemStore` is a map scan.

---
## Summary by cubic
`FSObjectStore.List` now walks only the subtree its prefix implies instead of the whole root, so filesystem backups no longer cost O(collections × all stored files). The walk runs inside `os.Root`, and results match a full-root prefix-filtered walk except that a prefix path through a symlink returns an error instead of an empty result.

**Behavior**
- The walk starts at the deepest directory the prefix fully names; a trailing partial segment stays a name filter, and a non-canonical prefix (leading `/`, `.` or `..` segment, doubled `/`) matches no key and yields nothing without touching disk.
- Every start-path component is Lstat'd — `os.Root` alone would follow relative in-root symlinks — and scoping stops before a segment containing a backslash, walking from its parent so Windows separator handling stays consistent.
- A missing prefix directory, a prefix naming a plain file, or a NUL byte returns nil; a missing or unreadable root remains an error.
- New tests compare scoped `List` against a full-root filtered walk and pin the symlink, scoping, non-canonical-prefix, and missing-root semantics.

<sup>Written for commit 3982d6ab52a3582dab4c12735a06b5d0691e7e5f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rostamlabs/rostam/pull/125?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved file listing with prefix-scoped traversal for more accurate results.
  * Prevented listings from accessing unrelated directories or escaping the configured root through symlinks.
  * Improved handling of missing directories, invalid prefixes, and file-based paths.
  * Missing storage roots now correctly return an error.

### CODE REVIEW

**Review by vahid-sohrabloo — COMMENTED** [review]
Reviewed at `7052d32`. The optimisation is right — it is the fix I asked for in the #111 review. One problem blocks it, and I reproduced it rather than reasoning about it, because it deletes files.

## Blocking: `List` can walk outside root, and `prune` then deletes what it finds

`listStart` checks containment with string comparison only (`filepath.Abs` + `HasPrefix`, no `EvalSymlinks`). `os.Stat` then follows symlinks, and `WalkDir` only `Lstat`s its own start argument. So a symlink as the LAST part of the path is safe, but a symlink in the MIDDLE is followed, and the walk ends up outside root. Keys are still built from the lexical path, so they look root-relative while the files are not.

With `<root>/acme` symlinked to an outside directory holding three snapshots, `List("acme/col/")`:

```
this PR: 3 keys  (acme/col/2020…snap, 2021…snap, 2022…snap)
main:    0 keys  (the old walk never follows the symlink)
```

Then `prune` deletes everything listed past retention, and `Delete` → `keyToPath` is lexical the same way, so `os.Remove` follows the link too. Same fixture, `retention=1`:

```
prune DELETED 2 files outside the backup root
```

On main those files cannot be found, so retention never touches them. This PR turns a quiet weakness into something that deletes data, which is why I think it should be fixed here and not filed.

The prefix is not the way in: `validateTenant` and `url.PathEscape` keep prefixes to two clean segments. What you need is a symlink inside `-backup-dir` — a normal operator move ("put one tenant on a bigger disk").

## Suggested fix

The module is on Go 1.26.5, so use `os.Root`: open the root once with `os.OpenRoot(f.root)` and walk `fs.WalkDir(r.FS(), startRel, …)`. The kernel refuses to step outside root, so the check stops being a string comparison. It also sorts out the error handling: `fs.ErrNotExist` means "no such prefix, return nil", anything else is a real error.

Worth saying in the comment what it does not cover: `os.Root` still follows symlinks that stay inside root, so `<root>/link -> <root>/acme` would produce `link/col/…` keys the old walk never produced. Harmless, but a difference.

Please add a symlink test — the fixture above is four lines, and there is none today.

The rest of the review is in my next comment.

**Review by vahid-sohrabloo — COMMENTED** [review]
The rest of the review. Good news first, because it decides whether the approach is right.

**Everything else matches the old behaviour.** A differential run compared this `List` against a plain full-root-walk-and-filter across **4,043 prefixes** (43 hand-picked — partial segments, trailing slashes, exact keys, empty, leading slash, `.`/`..`, a prefix naming a file, a partial segment matching a directory — plus 4,000 fuzzed), comparing key, size and mtime. **No differences** except the symlink case and the two below.

**The speedup is real.** 200 collections × 5 snapshots × 2 objects = 2,000 files. `List("acme/col0100/")`: **2.63 ms → 26.1 µs, about 100×**. With one collection (10 files) it is 41.5 µs vs 40.7 µs — no change, as expected. Worth landing.

## Two behaviour changes to decide on (both reproduced against main)

**1. A prefix whose directory part is not a directory now errors.** `os.IsNotExist` is the only `Stat` error you tolerate, so every other one fails the call. With key `a.snap` stored, `List("a.snap/x/y")`:

```
this PR: error — stat <root>/a.snap/x: not a directory
main:    no keys, no error
```

Same for `acme/col/<ts>.snap/x/y`, and for a NUL byte in a middle segment (`invalid argument`). Note `List` is the only method that never validates its input, while `keyToPath` rejects NUL. It also breaks the drop-in parity the file's own header promises: `MemStore.List` is a prefix filter and cannot fail this way.

**2. A missing root is now "empty" instead of an error.** Build the store, delete the root, then list:

```
this PR: no keys, no error
main:    error — lstat <root>: no such file or directory
```

This is the one that matters in production. `LatestKey` turns an empty list into `ErrNotFound`, so `RestoreLatest` would say "this collection has no snapshots" during a recovery when the real answer is that the backup volume is gone. `prune` would quietly prune nothing. Your own doc only promises nil for "a prefix whose directory does not exist yet", so keep the root an error: swallow `NotExist` only when `start != f.root`.

Both of these disappear under the `os.Root` walk, which is another reason to prefer it.

## Smaller points

- **The containment check in `listStart` is dead code.** `path.Clean("/"+prefix[:i])` already removed every `..` before the `Join`, so that branch can never run. Harmless — except its comment says escape is handled, and that is the belief that hid the symlink hole. Delete it and document `path.Clean` as the whole check (like `keyToPath` does), or make it a real check after resolving symlinks.

- **`TestFSObjectStoreListPrefixScopedWalk` does not test the scoping.** I replaced `listStart` with `return f.root` — optimisation gone — and it still passes. It is a good equivalence test, but nothing fails if the scoping is removed, so a later refactor could undo the 100× and stay green. Count `ReadDir` calls, or use a fixture where the unrelated subtree is big enough to time.

- **Case-insensitive filesystems.** `WalkDir` joins your start string with real entry names, so `List("ACME/col/")` on macOS or Windows returns `ACME/col/…` keys that match the prefix; the old walk returned `acme/col/…` and matched nothing. No caller re-cases a prefix and the Windows CI lane does not run `./backup/`, so one doc line is enough.

Net: right approach, the numbers justify it, equivalence is solid. Fix the symlink escape, and choose the two error changes on purpose rather than by accident.

**Comment by Adelagric** [comment]
Thanks — reproducing it through to `prune` deleting outside root was the right call, and the differential sweep plus the measurement are exactly what I would have wanted to run and did not.

`674757f` takes your design: `List` opens the root with `os.OpenRoot` and walks `fs.WalkDir(r.FS(), start)`, so containment is the kernel's check and the unreachable lexical check is gone (`path.Clean` is documented as the only lexical step, not relied on for containment). Your fixture, replayed: `List("acme/col/")` → `path escapes from parent`, `prune` fails on that, all three files outside root untouched.

One empirical point on the "residue": `r.FS()` refuses a start path with a symlink in *any* component, including one that stays inside root (`<root>/inlink → <root>/acme` gives the same `path escapes from parent`), so there is no symlink traversal at all — same as the full-root walk. A walk that merely *passes* a link still reports it as a plain entry, as before. Test covers both an escaping and an in-root link.

Error semantics, decided: (1) a prefix whose directory is missing, or names a plain file (`ENOTDIR`), or contains NUL → `nil`, matching `main` and `MemStore`'s pure prefix filter; (2) a missing root → error, as on `main` — `OpenRoot` gives that for free, test added; (3) a start path *through a symlink* → error, where `main` was silent. That last one is a deliberate change, by your own "operational teeth" argument: a tenant an operator symlinked onto another disk otherwise reads as "no snapshots" from `LatestKey`/`RestoreLatest` and as a silent no-op from `prune`. If you would rather keep silence there, it is one `errors.Is` away, but I think loud is right.

The scoping is now pinned: an unreadable directory planted in an unrelated subtree fails a root walk and must not fail `List("acme/col/")` — with `listStart` replaced by `return "."` that test fails. Case-insensitive filesystems documented on `List`. Measured here: 3.51 ms → 50 µs on your 2,000-file shape (`OpenRoot` per call included).

File: backup/fs.go [hunk 1]
```diff
@@ -157,18 +157,32 @@ func (f *FSObjectStore) Get(_ context.Context, key string) (io.ReadCloser, error
...
+	start := f.listStart(prefix)
+	if _, err := os.Stat(start); err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("fsstore: list %q: %w", prefix, err)
+	}
 	var out []objstore.ObjectInfo
-	err := filepath.WalkDir(f.root, func(p string, d fs.DirEntry, err error) error {
+	err := filepath.WalkDir(start, func(p string, d fs.DirEntry, err error) error {
```
[the remark this answers is not in the document]
Comment (Adelagric) [reply backup/fs.go]: Confirmed against `main` and fixed in `674757f` — the walk now runs inside `os.Root`, which refuses a start path with a symlink in any component; the maintainer's prune fixture no longer touches anything outside root. Regression test added.

File: backup/fs.go [hunk 2]
```diff
@@ -157,18 +157,32 @@ func (f *FSObjectStore) Get(_ context.Context, key string) (io.ReadCloser, error
...
+// single key, so walking the whole root would make each backup run cost
+// O(collections × all stored files). The key-prefix filter is still applied to
+// every entry, so the result is identical to a full-root walk.
+//
+// NOTE: like the S3 store, List BUFFERS every matching key in memory (one
+// ObjectInfo per matching file). This is bounded for the backup/retention use
+// (a handful of snapshots per collection prefix) but a prefix spanning millions
+// of files will materialize them all at once; scope the prefix narrowly.
 func (f *FSObjectStore) List(_ context.Context, prefix string) ([]objstore.ObjectInfo, error) {
+	start := f.listStart(prefix)
```
[the remark this answers is not in the document]
Comment (Adelagric) [reply backup/fs.go]: Fixed in `674757f`: a NUL prefix returns `nil` early (no key can contain NUL — `keyToPath` rejects it), and `ENOTDIR`/`ErrNotExist` on the start directory return `nil` too, matching `main` and `MemStore`. NUL and file-as-directory prefixes added to the equivalence test.

File: backup/fs.go [hunk 3]
```diff
@@ -157,38 +159,80 @@ func (f *FSObjectStore) Get(_ context.Context, key string) (io.ReadCloser, error
...
+		return nil, nil
+	}
+	r, err := os.OpenRoot(f.root)
+	if err != nil {
+		return nil, fmt.Errorf("fsstore: list %q: open root: %w", prefix, err)
+	}
+	defer func() { _ = r.Close() }()
+	fsys := r.FS()
+	start := listStart(prefix)
+	if _, err := fs.Stat(fsys, start); err != nil {
```
[the remark this answers is not in the document]
Comment (Adelagric) [reply backup/fs.go]: Confirmed — `os.Root` does follow a relative in-root link, and the previous fixture used an absolute target. Fixed in `b2cc16a`: every start-path component is Lstat'd through the `os.Root` handle and a symlink is an error; `os.Root` stays as the containment backstop for a link created between the Lstat and the walk. Test now covers a relative top-level link and a relative link below the tenant directory.

**Comment by Adelagric** [comment]
Correction to my previous comment, prompted by CodeRabbit's follow-up and confirmed empirically: `os.Root` refuses *absolute* and *root-escaping* symlink targets, but it **does** follow a relative symlink whose target stays inside root, and `fs.WalkDir` walks the target when its start is such a link. My "in-root" fixture used an absolute target, so it passed for the wrong reason — `List("rel/col/")` with `rel → acme` returned `rel/col/…` keys the full-root walk never produced. So your "residue" was right and my claim that there was none was wrong.

`b2cc16a` makes it actually no-follow: `listStart` Lstat's every component between root and the start directory through the `os.Root` handle — a symlink is the error we agreed on, a missing segment or a plain file yields nothing — with `os.Root` kept behind it as the containment backstop for a link created between the Lstat and the walk (worst case there is in-root aliasing, never escape). The symlink test now has a relative top-level link and a relative link below the tenant directory; disabling the check fails it.

Also from cubic: scoping stops at the first prefix segment containing a backslash and walks from its parent, so Windows — where `os.Root` would treat `\` as a separator while the key filter treats it literally — cannot return keys that do not carry the requested prefix. Unreachable via callers (`validateTenant` rejects `\`, collections are escaped), but free and equivalent everywhere.

File: backup/fs.go [hunk 4]
```diff
@@ -157,38 +159,80 @@ func (f *FSObjectStore) Get(_ context.Context, key string) (io.ReadCloser, error
...
+	if strings.ContainsRune(prefix, '\x00') {
+		return nil, nil
+	}
+	r, err := os.OpenRoot(f.root)
+	if err != nil {
+		return nil, fmt.Errorf("fsstore: list %q: open root: %w", prefix, err)
+	}
+	defer func() { _ = r.Close() }()
+	fsys := r.FS()
+	start := listStart(prefix)
```
[the remark this answers is not in the document]
Comment (Adelagric) [reply backup/fs.go]: Fixed in `b2cc16a`: scoping stops at the first prefix segment containing a backslash and walks from its parent, so the lexical key filter decides identically on every OS. Unreachable via callers (`validateTenant` rejects `\`, collection names are escaped), and a POSIX key with a literal backslash still lists — added to the equivalence test.

File: backup/fs.go [hunk 5]
```diff
@@ -201,6 +248,51 @@ func (f *FSObjectStore) List(_ context.Context, prefix string) ([]objstore.Objec
...
+	start = "."
+	i := strings.LastIndex(prefix, "/")
+	if i < 0 {
+		return start, true, nil
+	}
+	rel := strings.TrimPrefix(path.Clean("/"+prefix[:i]), "/")
+	if rel == "" {
+		return start, true, nil
+	}
+	for _, seg := range strings.Split(rel, "/") {
```
[the remark this answers is not in the document]
Comment (Adelagric) [reply backup/fs.go:280]: Fair — the walk was scoped by the cleaned form while the filter uses the raw one. Rather than fall back to a root walk: every key is a clean relative path, so a prefix whose directory part changes under `path.Clean` (leading `/`, `.`/`..` segment, doubled `/`) can match no key at all. `3982d6a` yields nothing for it without touching disk — same result as the full-root walk, zero I/O, and no chance of tripping on a symlink the raw prefix never named. Shapes added to the equivalence test; the symlink test checks `../link/col/` and `acme/../rel/col/` yield nothing rather than the link error.
